File: jinear-core/src/main/java/co/jinear/core/system/NormalizeHelper.java
package co.jinear.core.system;

import lombok.experimental.UtilityClass;

import java.io.UnsupportedEncodingException;
import java.text.Normalizer;
import java.util.Locale;

@UtilityClass
public class NormalizeHelper {

    private static final String LOCALE_EN = "en-EN";
    public static final String EMPTY_STRING = "";
    public static final String MASK_CHAR = "*";
    private static final String ACCENT_REGEX = "\\p{M}";
    private static final String ALPHANUMERIC_REGEX = "[^A-Za-z0-9]";
    private static final String ASCII_REGEX = "[^\\x00-\\x7F]";

    public static String removeAccent(String str) {
        return Normalizer.normalize(str.toLowerCase(), Normalizer.Form.NFD)
                .replaceAll(ACCENT_REGEX, EMPTY_STRING)
                .toLowerCase(new Locale(LOCALE_EN));
    }

    public static String removeNonAlphaNumeric(String str) {
        return Normalizer.normalize(str.toLowerCase(), Normalizer.Form.NFD)
                .replaceAll(ALPHANUMERIC_REGEX, EMPTY_STRING)
                .toLowerCase(new Locale(LOCALE_EN));
    }

    public static String removeNonAscii(String str) {
        return Normalizer.normalize(str.toLowerCase(), Normalizer.Form.NFD)
                .replaceAll(ASCII_REGEX, EMPTY_STRING)
                .toLowerCase(new Locale(LOCALE_EN));
    }

    public static String normalizeStrictly(String str) {
        return removeNonAlphaNumeric(removeNonAscii(removeAccent(str)));
    }

    public static String maskString(String str) {
        return maskBetweenChars(str, 0, str.length());
    }

    public static String maskBetweenChars(String str, int maskStart, int maskEnd) {
        String maskedPart = MASK_CHAR.repeat(maskEnd - maskStart);
        String preMaskedPart = str.substring(0, maskStart);
        String postMaskedPart = str.substring(maskEnd);
        return preMaskedPart + maskedPart + postMaskedPart;

    }

    public static void main(String[] args) throws UnsupportedEncodingException {
        String zort = "@012345";
        System.out.println(maskBetweenChars(zort, 2, 4));
        System.out.println(maskBetweenChars(zort, 0, zort.length()));
        System.out.println(normalizeStrictly(zort));
    }
}

File: jinear-web/utils/normalizeHelper.ts
const LOCALE_EN = "en-US";
const EMPTY_STRING = "";
const SPACE_STRING = " ";
const MASK_CHAR = "*";
const ACCENT_REGEX = /\\p{M}/;
const ALPHANUMERIC_REGEX = /[^A-Za-z0-9]/;
const USERNAME_REGEX = /[^A-Za-z0-9-_]/;
const ASCII_REGEX = /[\u{0080}-\u{10FFFF}]/gu;

const normalize = (str: string) => str?.normalize?.("NFD");

export const removeAccent = (str: string) =>
  normalize(str)
    ?.split(ACCENT_REGEX)
    ?.join(EMPTY_STRING)
    ?.toLocaleLowerCase?.(LOCALE_EN);

export const removeNonAlphaNumeric = (str: string) =>
  normalize(str)
    ?.split(ALPHANUMERIC_REGEX)
    ?.join(EMPTY_STRING)
    ?.toLocaleLowerCase?.(LOCALE_EN);

export const removeNonAscii = (str: string) =>
  normalize(str)
    ?.split(ASCII_REGEX)
    ?.join(EMPTY_STRING)
    ?.toLocaleLowerCase?.(LOCALE_EN);

export const removeUsernameNotAllowed = (str: string) =>
  normalize(str)
    ?.split(USERNAME_REGEX)
    ?.join(EMPTY_STRING)
    ?.toLocaleLowerCase?.(LOCALE_EN);

export const normalizeStrictly = (str: string) =>
  removeNonAlphaNumeric(removeNonAscii(removeAccent(str)));

export const normalizeUsername = (str: string) =>
  removeUsernameNotAllowed(removeNonAscii(removeAccent(str)));

export const normalizeUsernameReplaceSpaces = (str: string) =>
  normalizeUsername(str?.split(SPACE_STRING)?.join("-"));
